### PR TITLE
Update to arrow `19.0.0`

### PR DESCRIPTION
**Pull request by alamb**
~~Draft until  arrow 19.0.0 is released, see https://github.com/apache/arrow-rs/issues/2051 for details~~
Closes #2973 
Closes #2974 
Closes #2975

### CODE REVIEW

File: datafusion/common/src/scalar.rs [hunk 1]
```diff
@@ -1112,14 +1117,14 @@ impl ScalarValue {
         scalars: impl IntoIterator<Item = ScalarValue>,
         precision: &usize,
         scale: &usize,
-    ) -> Result<DecimalArray> {
+    ) -> Result<Decimal128Array> {
```
Comment (alamb) [inline datafusion/common/src/scalar.rs:1120]: `DecimalArray` now is `DecimalArray128` to add support of `DecimalArray256` -- in https://github.com/apache/arrow-rs/issues/2101

File: datafusion/core/src/avro_to_arrow/arrow_array_reader.rs [hunk 2]
```diff
@@ -130,58 +130,52 @@ impl<'a, R: Read> AvroArrowArrayReader<'a, R> {
         arrays.and_then(|arr| RecordBatch::try_new(projected_schema, arr).map(Some))
     }
 
-    fn build_boolean_array(
-        &self,
-        rows: RecordSlice,
-        col_name: &str,
-    ) -> ArrowResult<ArrayRef> {
+    fn build_boolean_array(&self, rows: RecordSlice, col_name: &str) -> ArrayRef {
```
Comment (alamb) [inline datafusion/core/src/avro_to_arrow/arrow_array_reader.rs:133]: This is another example showing the builders have become infallable

File: datafusion/physical-expr/src/expressions/cast.rs [hunk 3]
```diff
@@ -271,12 +271,18 @@ mod tests {
 
     #[test]
     fn test_cast_decimal_to_decimal() -> Result<()> {
-        let array = vec![1234, 2222, 3, 4000, 5000];
+        let array = vec![
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/cast.rs:274]: This was a bug in the test -- the input data had 5 rows but the expected value had 6 values (and apparently `is_valid()` used to return `false` for an index off the end of the array and in arrow 19 it returns `true`)

Comment (alamb) [reply datafusion/physical-expr/src/expressions/cast.rs:274]: cc @liu

**Review by tustvold — APPROVED** [review]

**Pull request merged**